### PR TITLE
Merge JITClientCompThread from the jitaas branch

### PR DESCRIPTION
**Pull request by a7ehuo**
Merge `JITClientCompThread` from the `jitaas` branch into the `master` branch.

Guard references to `TR_J9JITServerSharedCache`,
`TR_JITServerRelocationRuntime` with `JITSERVER_TODO`.

The mirrored change on the `jitaas` branch is PR #7232. 

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

### CODE REVIEW

**Comment by a7ehuo** [comment]
This PR has dependency on PR #7193.

**Comment by harryyu1994** [comment]
`TR_J9VMBase::getArrayLengthOfStaticAddress` was removed because it wasn't used anymore. We don't want to add it back in.

**Comment by harryyu1994** [comment]
Oh since this is a change going into `master`, you'll need to manually delete them. There are a couple of things that are already removed in `master` so we don't want to re-introduce them:
`bool TR::CompilationInfo::isRomClassForMethodInSharedCache(J9Method *method)`
`TR_YesNoMaybe TR::CompilationInfo::isMethodInSharedCache(J9Method *method)`
`   virtual bool getArrayLengthOfStaticAddress(void *ptr, int32_t &length);`

**Comment by a7ehuo** [comment]
@harryyu1994 Thanks for the heads up!. This PR depends on #7236. Once #7236 is finalized and merged, I'll update this PR to remove the following three unused methods: 

```
- bool TR::CompilationInfo::isRomClassForMethodInSharedCache(J9Method *method)
- TR_YesNoMaybe TR::CompilationInfo::isMethodInSharedCache(J9Method *method)
- virtual bool getArrayLengthOfStaticAddress(void *ptr, int32_t &length);
```

File: runtime/compiler/control/CompilationThread.cpp [hunk 1]
```diff
@@ -1479,6 +1482,17 @@ TR::CompilationInfo::disableAOTCompilations()
       }
 #endif
 
+#if defined(JITSERVER_SUPPORT)
+#if defined(J9VM_INTERP_AOT_RUNTIME_SUPPORT) && defined(J9VM_OPT_SHARED_CLASSES) && (defined(TR_HOST_X86) || defined(TR_HOST_POWER) || defined(TR_HOST_S390) || defined(TR_HOST_ARM) || defined(TR_HOST_ARM64))
+// Note: this method must be called only when we know that AOT mode for shared classes is enabled !!
+bool TR::CompilationInfo::isRomClassForMethodInSharedCache(J9Method *method)
```
[the remark this answers is not in the document]
Comment (a7ehuo) [reply runtime/compiler/control/CompilationThread.cpp]: Will remove the API and replace the call to `isRomClassForMethodInSharedCache` in `JITClientCompilationThread.cpp` 

```
isRomClassForMethodInSC = TR::CompilationInfo::get(fe->_jitConfig)->isRomClassForMethodInSharedCache(j9method);
```

to 
```
isRomClassForMethodInSC = TR::Options::sharedClassCache() ? fe->sharedCache()->isPointerInSharedCache(J9_CLASS_FROM_METHOD(j9method)->romClass) : false;
```

**Comment by a7ehuo** [comment]
@mpirvu All comments addressed. Ready for another review. Thanks.

**Comment by mpirvu** [comment]
Jenkins test sanity all jdk8,jdk11

**Comment by mpirvu** [comment]
On failure on mac Java8: `87 - - SCURLClassLoaderTests_SE80_1 `
```
Running command: /Users/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_mac_Personal/openjdkbinary/j2sdk-image/jre/bin/java -Xcompressedrefs -Xcompressedrefs -Xjit -Xgcpolicy:gencon    -Xshareclasses:persistent,controlDir=.,name=FindStore,noTimestampChecks -cp ./FindStore/E.jar:./FindStore/Nothing.jar:./FindStore/F.jar:./FindStore/G.jar:./FindStore/H.jar jnurlcldr.shared.findstore.E_Main
Time spent starting: 3 milliseconds
Time spent executing: 2762 milliseconds
Test result: FAILED
 [OUT] CACHED E
 [OUT] CACHED F
 [OUT] CACHED G
 [OUT] CACHED H
 [OUT] Result=3
 [ERR] Unhandled exception
 [ERR] Type=Segmentation error vmState=0x0005ffff
 [ERR] J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
 [ERR] Handler1=0000000002CA1F00 Handler2=00000000022D5C20 InaccessibleAddress=0000000000010000
 [ERR] RDI=00000000040A95A8 RSI=00007000002DC384 RAX=0000000000000000 RBX=00000000040A95A8
 [ERR] RCX=000000000000004A RDX=00007FA69BF0A220 R8=00000000FFFFFFF3 R9=000000000000000D
 [ERR] R10=00007FA69BF0A4A0 R11=0000000000010000 R12=00000000FFFFFFF3 R13=00007000002DC9E8
 [ERR] R14=000000000000004A R15=0000000000000054
 [ERR] RIP=00007FFF9E9BB7D4 GS=0000 FS=0000 RSP=00007000002DC350
 [ERR] RFlags=0000000000010202 CS=002B RBP=00007000002DC370 ERR=0001000000000004
 [ERR] TRAPNO=000000040000000E CPU=0000000000040000 FAULTVADDR=0000000000010000
 [ERR] XMM0 0000000000000000 (f: 0.000000, d: 0.000000e+00)
 [ERR] XMM1 0000f80100000014 (f: 20.000000, d: 1.347234e-309)
 [ERR] XMM2 000007d000000fa0 (f: 4000.000000, d: 4.243992e-311)
 [ERR] XMM3 0000000000000000 (f: 0.000000, d: 0.000000e+00)
 [ERR] XMM4 0000009600000320 (f: 800.000000, d: 3.182994e-312)
 [ERR] XMM5 ffffffffffffffff (f: 4294967296.000000, d: nan)
 [ERR] XMM6 0000000100000001 (f: 1.000000, d: 2.121996e-314)
 [ERR] XMM7 8000001580000015 (f: 2147483648.000000, d: -4.562291e-313)
 [ERR] XMM8 8000000080000000 (f: 2147483648.000000, d: -1.060998e-314)
 [ERR] XMM9 0000000000000000 (f: 0.000000, d: 0.000000e+00)
 [ERR] XMM10 0000000000000000 (f: 0.000000, d: 0.000000e+00)
 [ERR] XMM11 0000000000000000 (f: 0.000000, d: 0.000000e+00)
 [ERR] XMM12 0000000000000000 (f: 0.000000, d: 0.000000e+00)
 [ERR] XMM13 0000000000000000 (f: 0.000000, d: 0.000000e+00)
 [ERR] XMM14 0000000000000000 (f: 0.000000, d: 0.000000e+00)
 [ERR] XMM15 0000000000000000 (f: 0.000000, d: 0.000000e+00)
 [ERR] Module=/usr/lib/system/libsystem_c.dylib
 [ERR] Module_base_address=00007FFF9E95C000 Symbol=__findenv
 [ERR] Symbol_address=00007FFF9E9BB780
 [ERR] 
 [ERR] Method_being_compiled=java/lang/String.indexOf(I)I
 [ERR] Target=2_90_20191007_179 (Mac OS X 10.11.6)
 [ERR] CPU=amd64 (4 logical CPUs) (0x200000000 RAM)
 [ERR] ----------- Stack Backtrace -----------
 [ERR] ---------------------------------------
 [ERR] JVMDUMP039I Processing dump event "gpf", detail "" at 2019/10/07 17:17:07 - please wait.
 [ERR] JVMDUMP032I JVM requested System dump using '/Users/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_mac_Personal/openjdk-tests/TestConfig/test_output_15704818448005/SCURLClassLoaderTests_SE80_1/core.20191007.171707.70206.0001.dmp' in response to an event
 [ERR] JVMDUMP010I System dump written to /Users/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-64_mac_Personal/openjdk-tests/TestConfig/test_output_15704818448005/SCURLClassLoaderTests_SE80_1/core.20191007.171707.70206.0001.dmp
```

**Comment by mpirvu** [comment]
The error on mac happens either just before compilation or just after the compilation has ended, so it is suspicious.

**Comment by mpirvu** [comment]
Jenkins test sanity osx jdk8

**Comment by mpirvu** [comment]
If this PR introduced the problem on mac it could only have come from the implementation of `TR::CompilationInfo::canRelocateMethod(TR::Compilation *comp)`
@a7ehuo is this new method used by the non-jitaas code?

**Comment by a7ehuo** [comment]
`canRelocateMethod()` is only used in the `JITServer` mode in `JITClientCompilationThread.cpp`. Its declaration and definition are also guarded by `JITSERVER_SUPPORT`. 

I'm also trying to run `SCURLClassLoaderTests_SE80_1` in a Grinder: https://ci.adoptopenjdk.net/job/Grinder/753

**Comment by mpirvu** [comment]
I will wait for the mac test to finish, but at this point I highly doubt the error is due to this PR

**Comment by mpirvu** [comment]
All tests have passed, hence merging.

**Pull request merged**